engine/src/Interpreter/interpreter_ops.cuh

```diff
@@ -677,253 +391,358 @@ private:
+					setColumnValid(row_valids, output_position, !!left_value);
+				} else if(right_valid) {
+					store_data_in_buffer(right_value, buffer, output_position);
+					setColumnValid(row_valids, output_position, !!right_valid);
+				}	else {
+					setColumnValid(row_valids, output_position, false);
 				}
 			} else if(oper == BLZ_MAGIC_IF_NOT) {
 				if(left_valid && left_value) {
-					store_data_in_buffer<OutputTypeOperator>(right_value, buffer, output_position);
+					store_data_in_buffer(right_value, buffer, output_position);
 					setColumnValid(row_valids, output_position, right_valid);
 				} else {
 					// we want to indicate to first_non_magic to use the second value
-					store_data_in_buffer<OutputTypeOperator>(
-						getMagicNumber<OutputTypeOperator>(), buffer, output_position);
+					store_data_in_buffer(
+						getMagicNumber<RightType>(), buffer, output_position);
 				}
 			} else if(oper == BLZ_FIRST_NON_MAGIC) {
 				if(left_value == getMagicNumber<LeftType>()) {
-					store_data_in_buffer<OutputTypeOperator>(right_value, buffer, output_position);
+					store_data_in_buffer(right_value, buffer, output_position);
 					setColumnValid(row_valids, output_position, right_valid);
 				} else {
-					store_data_in_buffer<OutputTypeOperator>(left_value, buffer, output_position);
+					store_data_in_buffer(left_value, buffer, output_position);
 					setColumnValid(row_valids, output_position, left_valid);
 				}
-			} else if(oper == BLZ_STR_LIKE || oper == BLZ_STR_SUBSTRING || oper == BLZ_STR_CONCAT) {
-				store_data_in_buffer<OutputTypeOperator>(left_value, buffer, output_position);
-				setColumnValid(row_valids, output_position, left_valid);
+			} else {
+				setColumnValid(row_valids, output_position, false);
 			}
 		} else {
-			// It's a unary operation			
+			// It's a unary operation, scalar inputs are not allowed
+			assert(left_position >= 0);
 
 			LeftType left_value;
-			bool left_valid;
-			if(left_position >= 0) {
-				left_value = get_data_from_buffer<LeftType>(buffer, left_position);
-				left_valid = getColumnValid(row_valids, left_position);
-			} else if(left_position == SCALAR_INDEX) {
-				left_value = ((LeftType *) this->scalars_left)[op_index];
-				left_valid = true;
-			} else if(left_position == SCALAR_NULL_INDEX) {
-				left_valid = false;
-			}
-
-			OutputTypeOperator computed = left_value;
-
-			gdf_unary_operator oper = this->unary_operations[op_index];
+			get_data_from_buffer(&left_value, buffer, left_position);
+			bool left_valid = getColumnValid(row_valids, left_position);
+			
+			gdf_unary_operator oper = unary_operations[op_index];
 			if (left_valid) {
 				if(oper == BLZ_FLOOR) {
-					computed = floor(left_value);
+					double val = static_cast<double>(left_value);
+					store_data_in_buffer(floor(val), buffer, output_position);
 				} else if(oper == BLZ_CEIL) {
-					computed = ceil(left_value);
+					double val = static_cast<double>(left_value);
```

Comment:
the previous code seems easier to follow most operations are just 
```
computed = operation(left_value);
```

Reply:
same  as above just making it explicit

Reply:
this is more lines of code and both do the same thing.